pkg/generator/generator.go

```diff
@@ -53,205 +57,239 @@ type keyGenConfig struct {
+			break
+		}
+		// make queue from requeue
+		queue := append(requeue[:0:0], requeue...)
+		// reset re-work queue
+		requeue = []*v1alpha1.KeyConfig{}
+		// process all keys in queue
+		for _, key := range queue {
+			log := g.Log.WithValues(
 				"data_key", key.Name,
 				"secret_type", string(key.Type))
-			return err
-		}
-		empty, err := keyGenerator.SecretManagerHasData(ctx)
-		if err != nil {
-			return err
-		}
-		if empty {
-			// load dependencies.
-			// missing secrets and keys are treated are errors and should stop generation
-			// self referencing deps can be resolved by re-running again later
-			reEnqueForKey, err := keyGenerator.ConfigureDependencies(ctx, rclient)
-			// check for NotFoundErr, for retry
+			log.V(1).Info("key started")
+			keyGenerator := newKeyGenerator(key, g)
+			empty, err := keyGenerator.secretManagerHasData(ctx)
 			if err != nil {
+				log.Error(err, "skipping key")
 				return err
 			}
-			// add key to queue if the missing dependency is in the queue
-			if err := enqueuer(reEnqueuForKey, key); err != nil {
+			// There's no secret data after checking, so get dependencies
+			if empty {
+				log.V(0).Info("secret needs to be generated")
+				log.V(1).Info("gathering dependencies")
+				// load dependencies.
+				// missing secrets and keys are treated are errors and should stop generation
+				// self referencing deps can be resolved by re-running again later
+				reEnqueForKey, err := keyGenerator.configureDependencies(ctx)
+				if k8serror.IsNotFound(err) {
+					log.V(0).Info("has unmet dependencies will retry")
+					return err
+				}
+				// this is an actual error
+				if err != nil {
+					log.Error(err, "error during dependecy resolution will retry")
+					return err
+				}
+				// if this is empty all deps are met, otherwise try another key
+				if reEnqueForKey != "" {
+					log.V(1).Info("trying to resolve dependency without a retry")
+					requeue, err = enquerer(requeue, key, reEnqueForKey)
+					continue
+				}
+				// add key to queue if the missing dependency is in the queue
+				//	enqueuer(reEnqueuForKey, key); err != nil {
 				// not found
+			}
+			// Ensure Secret Manager and Secret Object are in a generated state
+			if err := keyGenerator.updateSecretKeys(ctx); err != nil {
+				// return with a retry
+				log.Error(err, "could't generate to secret manager, skipping")
 				return err
 			}
+			log.V(1).Info("key completed")
 		}
-		if err := keyGenerator.GenAndUpdateSecretMgr(); err != nil {
-			// return with a retry
-			return err
-		}
-		return nil
-	} // end of keys loop
+	}
+	g.Log.V(1).Info("completed all keys")
+	return nil
 }
 
 func newKeyGenerator(
-	key *v1alpha1.KeyConfig, genConfig *GenConfig) (*keyGenConfig, error) {
+	key *v1alpha1.KeyConfig,
+	genConfig *GenConfig) *keyGenConfig {
 
 	var keyInterface KeyMgr
 	switch key.Type {
 	case v1alpha1.KeyConfigTypeCA:
-		keyInterface = NewRootCA(key)
+		keyInterface, _ = NewRootCA(key)
```

Comment:
if we always ignore the error, should we change the return types?

Reply:
How could this fail ? If the NewXX can never fail, then perhaps err is not needed

Reply:
There's a couple key types that could fail. e.g. we create a temporary directory for a keystore on init so all work is done in a single dir and then cleaned.

Reply:
I think you had brought that up before. We were catching them.

Reply:
I think you didn't get why creating an object would have a err.

Reply:
only two initializers now throw errors